wormhole/aggregator/src/circuits/tree.rs

```diff
@@ -0,0 +1,221 @@
+use plonky2::{
+    field::extension::Extendable,
+    hash::hash_types::RichField,
+    iop::witness::{PartialWitness, WitnessWrite},
+    plonk::{
+        circuit_builder::CircuitBuilder,
+        circuit_data::{CircuitData, CommonCircuitData, VerifierOnlyCircuitData},
+        config::GenericConfig,
+    },
+};
+use rayon::{iter::ParallelIterator, slice::ParallelSlice};
+use wormhole_verifier::ProofWithPublicInputs;
+use zk_circuits_common::circuit::{C, D, F};
+
+use crate::TREE_BRANCHING_FACTOR;
+
+/// A proof containing both the proof data and the circuit data needed to verify it.
+#[derive(Debug)]
+pub struct AggregatedProof<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
+{
+    pub proof: ProofWithPublicInputs<F, C, D>,
+    pub circuit_data: CircuitData<F, C, D>,
+}
+
+pub fn aggregate_to_tree(
+    leaf_proofs: Vec<ProofWithPublicInputs<F, C, D>>,
+    common_data: &CommonCircuitData<F, D>,
+    verifier_data: &VerifierOnlyCircuitData<C, D>,
+) -> anyhow::Result<AggregatedProof<F, C, D>> {
+    // Aggregate the first level.
+    let mut proofs = aggregate_level(leaf_proofs, common_data, verifier_data)?;
+
+    // Do the next levels by utilizing the circuit data within each aggregated proof.
+    while proofs.len() > 1 {
+        let common_data = &proofs[0].circuit_data.common.clone();
+        let verifier_data = &proofs[0].circuit_data.verifier_only.clone();
+        let to_aggregate = proofs.into_iter().map(|p| p.proof).collect();
+
+        let aggregated_proofs = aggregate_level(to_aggregate, common_data, verifier_data)?;
+
+        proofs = aggregated_proofs;
+    }
+
+    assert!(proofs.len() == 1);
+    Ok(proofs.pop().unwrap())
+}
+
+#[cfg(not(feature = "multithread"))]
+fn aggregate_level(
+    proofs: Vec<ProofWithPublicInputs<F, C, D>>,
+    common_data: &CommonCircuitData<F, D>,
+    verifier_data: &VerifierOnlyCircuitData<C, D>,
+) -> anyhow::Result<Vec<AggregatedProof<F, C, D>>> {
+    let mut aggregated_proofs = Vec::with_capacity(proofs.len() / TREE_BRANCHING_FACTOR);
+
+    for pair in proofs.chunks(TREE_BRANCHING_FACTOR) {
+        let proof_a = &pair[0];
+        let proof_b = &pair[1];
+
+        let aggregated_proof = aggregate_pair(proof_a, proof_b, common_data, verifier_data)?;
+        aggregated_proofs.push(aggregated_proof);
+    }
+
+    Ok(aggregated_proofs)
+}
+
+#[cfg(feature = "multithread")]
+fn aggregate_level(
+    proofs: Vec<ProofWithPublicInputs<F, C, D>>,
+    common_data: &CommonCircuitData<F, D>,
+    verifier_data: &VerifierOnlyCircuitData<C, D>,
+) -> anyhow::Result<Vec<AggregatedProof<F, C, D>>> {
+    proofs
+        .par_chunks(TREE_BRANCHING_FACTOR)
+        .map(|pair| {
+            let proof_a = &pair[0];
+            let proof_b = &pair[1];
+
+            aggregate_pair(proof_a, proof_b, common_data, verifier_data)
```

Comment:
it seems like this only works for TREE_BRANCHING_FACTOR = 2, can we make it handle other values?